File: exercises/RB101_small_problems/medium_2/unlucky_days.rb
=begin
# --------------------------- Problem
  Restate the problem:

  Write a method that returns the number of Friday the 13ths
  in a year given by an argument. Assume the year is greater than
  1752 (when the United Kingdom adopted the modern Gregorian Calendar)
  and that it will remain in use in for the foreseeable future.

  Input: integer (year)
  Output: integer (num of days)

# --------------------------- Test Cases + Rules

  Explicit:
    - assume year is greater 1752
    - return number of friday the 13ths in given year

  Implicit:
    - none identified

# --------------------------- Data Structure

# --------------------------- Scrapbook

  - initialize Date object
  - look at #friday? method

# --------------------------- Algorithm
  /* given an integer representing the year */
    - initialize date to new date object
    - initialize count variable
    - iterate 12 times over date
      - if date is friday? increment count
    - return count

=end

require 'date'

def friday_13th(year)
  date = Date.new(year, 12, 13)
  count = 0

  11.times do
    count += 1 if date.friday?
    date = date << 1 
    # ^ reassing date to 1 month before self
  end

  count
end

p friday_13th(2015) == 3
p friday_13th(1986) == 1
p friday_13th(2019) == 2

File: exercises/JS210_Small_Problems/Medium_2/unluckyDays.js
/*
---------- PROBLEM ----------
INPUT:
  - integer (year)
OUTPUT:
  - integer (number of fridays on the 13th day of the month)

EXAMPLES:

----------- RULES -----------
  - Assume that the year is greater than 1752
    - Assume the same calendar for future years
  - One arg will be provided in integer form representing the year

------- DATA STRUCTURE ------
  - Date, String

--------- SCRATCH PAD -------

---------- ALGORITHM --------
  - Iterate over all the months of the given year.
  - For each month, get the day that falls on the 13th.
  - Count the 13ths that fall on a Friday.
  - Return the count.
*/

function fridayThe13ths(year) {
  let count = 0;

  for (let month = 0; month < 12; month++) {
    const date = new Date(year, month, 13);
    const dayOfWeek = date.getDay();
    // friday would be 5
    if (dayOfWeek === 5) count += 1;
  }

  return count;
}

console.log(fridayThe13ths(1986));      // 1
console.log(fridayThe13ths(2015));      // 3
console.log(fridayThe13ths(2017));      // 2